### PR TITLE
reduced the redundant code in reduce [refs #255]

### PR DESCRIPTION
**Pull request by palashahuja**

### CODE REVIEW

**Comment by palashahuja** [comment]
@ankurankan Please let me know any changes in the name of variables that have to be made. Also, the above commit removes bug #222 .
Also,
- There are no pep8 errors for now.
- I couldn't think of a better way to delete elements from a dictionary other than using a `for` loop.
- If there is one element that is to be reduced, then what should `factor.values` be replaced by? Although, I have assigned it an empty `np.array([],dtype = np.int32)`

**Comment by palashahuja** [comment]
Also in the 
`if not all((itertools.starmap(lt,zip(value_indices,phi.cardinality[reduced_indices]))))` statement takes around `7.56us` while `if not all(value_indices < phi.cardinality[reduced_indices])` takes `10.4us`, although the former is bit longer and is quite difficult to read. What should I use ? .

**Comment by ankurankan** [comment]
@palashahuja: About #222, I have updated the issue with the expected output. Please fix your code accordingly.

File: pgmpy/factors/Factor.py [hunk 1]
```diff
@@ -261,28 +262,26 @@ def reduce(self, values, inplace=True):
...
+            
+        if not all('_' in value for value in values):
+            raise TypeError("Values should be in the form of variablename_index")
+        value_row = list(zip(*[value.split('_') for value in values]))
+        reduced_variables = list(value_row[0])
+        value_indices = list(map(int, value_row[1]))
+        if not set(reduced_variables).issubset(set(factor.scope())):
+            raise Exceptions.ScopeError("%s not in scope" % list(set(reduced_variables)-set(factor.scope())))
+        reduced_indices = np.where(np.in1d(factor.scope(), reduced_variables))
+        if not all(itertools.starmap(lt, zip(value_indices, (factor.cardinality)[reduced_indices]))):
```
Comment (ankurankan) [inline pgmpy/factors/Factor.py]: Redundant parenthesis in `(factor.cardinality)`

**Comment by ankurankan** [comment]
@palashahuja Also add a few blank lines in between to separate parts of code. Presently it looks too jumbled up.

**Comment by palashahuja** [comment]
@ankurankan Please let me know about all the changes. So that I can stash all the changes in one single commit.

**Comment by ankurankan** [comment]
@palashahuja And I think you can go with using `itertools.starmap`, it's not that complex and is save a good amount of time. 

Make these changes and I think the PR should be good to go.

**Comment by palashahuja** [comment]
I have already fixed the code though but I am facing the following conundrum :

```
In [6]:phi = Factor(['x1','x2','x3'],[2,3,2],np.arange(12))
In [7]:phi.reduce(['x1_0','x2_1','x3_1'])
In [8]:print(phi)
x1      x2      x3      phi(x1, x2, x3)
------------------------------------------------------
x1_0        x2_0        x3_0        3.0000
```

although

```
In [9]:phi.variables
OrderedDict([('x1', ['x1_0']), ('x2', ['x2_1']), ('x3', ['x3_1'])])
In [10]:phi.cardinality
array([1, 1, 1])
```

Am I doing something wrong over here or is that what you are exactly expecting ?

**Comment by abinashpanda** [comment]
@palashahuja This is the correct behaviour.

**Comment by ankurankan** [comment]
@palashahuja The behavior of reduce is correct but there's a bug in `__str__` function #302

**Comment by palashahuja** [comment]
@ankurankan @abinashpanda  any further changes , please have a look.

**Comment by ankurankan** [comment]
@palashahuja Could you please also add a few more tests for `reduce` method?

**Comment by palashahuja** [comment]
@ankurankan Sure no problem...

File: pgmpy/factors/Factor.py [hunk 2]
```diff
@@ -261,27 +262,31 @@ def reduce(self, values, inplace=True):
...
+        if not set(reduced_variables).issubset(set(factor.scope())):
+            raise Exceptions.ScopeError('%s not in scope' % list(set(reduced_variables)-set(factor.scope())))
+
+        reduced_indices = np.where(np.in1d(factor.scope(), reduced_variables))
+        if not all(itertools.starmap(lt, zip(value_indices, factor.cardinality[reduced_indices]))):
+            raise Exceptions.SizeError('Value is greater than max possible value')
+
+        reduce_assign = np.full(len(factor.cardinality), -1)
+        reduce_assign[reduced_indices] = value_indices
+        factor.values = factor.values[list(map(int,factor._index_for_assignment(reduce_assign)))]
```
Comment (abinashpanda) [inline pgmpy/factors/Factor.py:281]: PEP8 space after `,`. @palashahuja please use pep8 checker before submitting a PR.

**Pull request merged**